### PR TITLE
Support for tracking stack traces

### PR DESCRIPTION
**Pull request by ahigerd**
In order to support using mGBA (and medusa) for debugging and reverse engineering, this PR adds functionality to track calls and returns, inspect the call stack, and break execution on function boundaries.

This PR only provides an ARM implementation, and only for the CLI debugger. Future enhancements can add support for other debuggers and the SM83 core, as well as additional commands such as step-over and break-on-exception.

### CODE REVIEW

File: src/arm/debugger/debugger.c [hunk 1]
```diff
@@ -133,6 +246,8 @@ void ARMDebuggerDeinit(struct mDebuggerPlatform* platform) {
 	}
 	ARMDebugBreakpointListDeinit(&debugger->swBreakpoints);
 	mWatchpointListDeinit(&debugger->watchpoints);
+
```
Comment (endrift) [inline src/arm/debugger/debugger.c:244]: Nit: This is a block of deinits, so this newline is extraneous.

File: src/arm/debugger/debugger.c [hunk 2]
```diff
@@ -320,12 +435,22 @@ static void ARMDebuggerTrace(struct mDebuggerPlatform* d, char* out, size_t* len
...
 
-	*length = snprintf(out, *length, "%08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X cpsr: %08X | %s",
-		               cpu->gprs[0],  cpu->gprs[1],  cpu->gprs[2],  cpu->gprs[3],
-		               cpu->gprs[4],  cpu->gprs[5],  cpu->gprs[6],  cpu->gprs[7],
-		               cpu->gprs[8],  cpu->gprs[9],  cpu->gprs[10], cpu->gprs[11],
-		               cpu->gprs[12], cpu->gprs[13], cpu->gprs[14], cpu->gprs[15],
-		               cpu->cpsr.packed, disassembly);
+	size_t regStringLen = *length;
+	ARMDebuggerFormatRegisters(&cpu->regs, out, &regStringLen);
+	*length = regStringLen + snprintf(out + regStringLen, *length - regStringLen, " | %s", disassembly);
```
Comment (endrift) [inline src/arm/debugger/debugger.c:435]: Nit: I'd prefer this split onto two lines.

File: src/arm/debugger/debugger.c [hunk 3]
```diff
@@ -320,12 +435,22 @@ static void ARMDebuggerTrace(struct mDebuggerPlatform* d, char* out, size_t* len
...
+	*length = snprintf(out, *length, "%08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X %08X cpsr: %08X",
+		               regs->gprs[0],  regs->gprs[1],  regs->gprs[2],  regs->gprs[3],
+		               regs->gprs[4],  regs->gprs[5],  regs->gprs[6],  regs->gprs[7],
+		               regs->gprs[8],  regs->gprs[9],  regs->gprs[10], regs->gprs[11],
+		               regs->gprs[12], regs->gprs[13], regs->gprs[14], regs->gprs[15],
+		               regs->cpsr.packed);
+}
+
+static void ARMDebuggerFrameFormatRegisters(struct mStackFrame* frame, char* out, size_t* length) {
+	ARMDebuggerFormatRegisters((struct ARMRegisterFile*)frame->regs, out, length);
```
Comment (endrift) [inline src/arm/debugger/debugger.c:448]: Casting from `void*` is implicit in C, so this cast isn't needed. (And if it were, mGBA style says there should be a space between the cast and the variable.)

File: src/debugger/cli-debugger.c [hunk 4]
```diff
@@ -159,8 +178,12 @@ static void _continue(struct CLIDebugger* debugger, struct CLIDebugVector* dv) {
 }
 
 static void _next(struct CLIDebugger* debugger, struct CLIDebugVector* dv) {
+	struct mDebuggerPlatform* platform = debugger->d.platform;
```
Comment (endrift) [inline src/debugger/cli-debugger.c:181]: Nit: Variable declarations go after `UNUSED`s in mGBA style.

File: src/debugger/stack-trace.c [hunk 5]
```diff
@@ -0,0 +1,100 @@
+/* Copyright (c) 2013-2016 Jeffrey Pfau
```
Comment (endrift) [inline src/debugger/stack-trace.c:1]: It's also not 2016 anymore.

File: src/debugger/stack-trace.c [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+size_t mStackTraceGetDepth(struct mStackTrace* stack) {
+	return mStackFramesSize(&stack->stack);
+}
+
+struct mStackFrame* mStackTracePush(struct mStackTrace* stack, uint32_t pc, uint32_t destAddress, uint32_t sp, void* regs) {
+	struct mStackFrame* frame = mStackFramesAppend(&stack->stack);
+	frame->callAddress = pc;
+	frame->entryAddress = destAddress;
+	frame->frameBaseAddress = sp;
+	frame->regs = malloc(stack->registersSize);
```
Comment (endrift) [inline src/debugger/stack-trace.c:46]: The `malloc`/`free` during runtime makes me sad, but unfortunately mGBA doesn't have an appropriate utility type for this yet. That can be a later enhancement I suppose.

File: include/mgba/internal/debugger/stack-trace.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
+/* Copyright (c) 2013-2019 Jeffrey Pfau
```
[the remark this answers is not in the document]
Comment (ahigerd) [reply include/mgba/internal/debugger/stack-trace.h:1]: Don't you wish it were?

[the remark this answers is not in the document]
Comment (endrift) [reply include/mgba/internal/debugger/stack-trace.h:1]: Why stop there?

**Pull request merged**